File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    private int searchLastPos(int[] nums, int target, int firstPos){
        int left = firstPos;
        int right = nums.length - 1;
        int mid = 0;
        
        while (left <= right){
            mid = (right + left) / 2;
            
            if (nums[mid] == target){
                if (mid == nums.length - 1 
                    || nums[mid + 1] != nums[mid]){
                    return mid;
                }
                else {
                    left = mid + 1;
                }
            }
            
            if (nums[mid] <= target){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }
        
        return -1;
    }
    
    public int[] searchRange(int[] nums, int target) {
        int[] result = {-1,-1};
        
        int left = 0;
        int right = nums.length - 1;
        int mid = 0;
        
        while(left <= right){
            mid = (right + left) / 2;
            
            if (nums[mid] == target) {
                if (mid == 0 || nums[mid - 1] != nums[mid]){
                    result[0] = mid;
                    result[1] = searchLastPos(nums, target, mid);
                    break;
                }
                else{
                    right = mid - 1;
                }
            }
            
            if (nums[mid] < target){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }
        
        return result;
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    std::vector<int> searchRange(std::vector<int>& nums, int target) {
        if (nums.size() == 0) {
            return { -1, -1 };
        }

        int left = 0;
        int right = nums.size() - 1;
        std::vector<int> res = { -1, -1 };
        
        if (nums[left] == target) {
            res[0] = left;
        }
        if (nums[right] == target) {
            res[1] = right;
        }

        while (left + 1 < right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                if (nums[mid - 1] != target) {
                    res[0] = mid;
                    right = nums.size() - 1;
                }
                if (nums[mid + 1] != target) {
                    res[1] = mid;
                    left = 0;
                }

                if (res[1] == -1) {
                    left = mid;
                }
                else if (res[0] == -1) {
                    right = mid;
                }
                else {
                    return res;
                }
            }
            else {
                if (nums[mid] < target) {
                    left = mid;
                }
                else {
                    right = mid;
                }
            }
        }

        if (nums[left] == target || nums[right] == target) {
            if (res[0] == -1) {
                res[0] = nums[left] == target ? left : right;
            }
            if (res[1] == -1) {
                res[1] = nums[right] == target ? right : left;
            }
        }

        return res;
    }
};